ocamldoc: use ocamldoc.opt when available to build documentation and manpages

gasche opened the pull request:
This PR tweaks the build system of ocamldoc to use ocamldoc.opt to build documentation when available, and it also tweaks the root Makefile to build manpages (whose compilation time has been a hot topic of discussion in #8835) later in the build, when ocamldoc.opt may be available.

Before that PR, building the manpages takes 14s on my machine (it was 16s before #8836). After that PR, it takes 4s (5s before #8836). All other document targets benefit (`html_doc`, introduced to test HTML rendering of documentation comments, goes from 36s to 8.5s) but they are not part of the default build.

To check this PR I tried to run all the tests from ocamldoc's Makefile; it looks like some of them have bitrotten away, and I removed one that could not be run at all.

gasche committed:
ocamldoc/Makefile: manpages don't need generators anymore, odoc_man is builtin

gasche committed:
ocamldoc: build manpages explicitly in root Makefile

gasche committed:
ocamldoc/Makefile: use ocamldoc.opt when available

gasche committed:
ocamldoc: remove an obsolete test

The odoc_fhtml file required by this test seems to have been part of
the ocamldoc distribution sometime in the distant past (see
27934ab79a7ee3a62e78c3bfea489d0def8ff7e2 from 2002) but I can't find
traces of it in the git history of the compiler distribution.

dra27 commented:
I've never been sure why opam-repository does this, but the `ocaml-base-compiler` packages all do `make world` followed by `make world.opt` - as they stand, they'd fail to take advantage of this change, therefore (hopefully they've just done it that by accident).

dra27 commented on ocamldoc/Makefile:
```diff
@@ -44,24 +44,25 @@ OCAMLPP=-pp 'sh ./remove_DEBUG'
 OCAMLDOC=ocamldoc
+OCAMLDOC_OPT=$(OCAMLDOC).opt
 
-# TODO: clarify whether the following really needs to be that complicated
```
It may still be the case that it can be less complicated, but it is necessary!

gasche replied:
Thanks for the explanation. I killed the commit that implements the wrong simplification, but I may try to write a small comment for why the various cases are necessary.

dra27 commented on ocamldoc/Makefile:
```diff
@@ -44,24 +44,25 @@ OCAMLPP=-pp 'sh ./remove_DEBUG'
-  else
-    OCAMLDOC_RUN=$(OCAMLRUN) ./$(OCAMLDOC)
-  endif
+  OCAMLDOC_RUN_BYTE=$(OCAMLRUN) -I $(ROOTDIR)/otherlibs/$(UNIXLIB) -I $(ROOTDIR)/otherlibs/str ./$(OCAMLDOC)
```
This breaks `make world` on a `--disable-shared` build. The issue is that when shared support isn't available, `ocamldoc` is built with `-custom` so mustn't be invoked via `ocamlrun`

gasche replied:
Note: I can't find where in the Makefile logic `-custom` is used in absence of shared libraries (or at all) to compile ocamldoc. (There is logic to that effect in `Makefile.tools` but as far as I can tell it's not used to build ocamldoc.)

dra27 replied:
It's done using the "lib_custom" flag in and `str.cmo` & `unix.cmo` - on a system without shared linking, using the `.cmo` forces `-custom` mode (see https://github.com/ocaml/ocaml/blob/trunk/file_formats/cmo_format.mli#L55 and https://github.com/ocaml/ocaml/blob/trunk/bytecomp/bytelink.ml#L55)

gasche replied:
Wow. Thanks. I'll definitely add a comment in the ocamldoc Makefile to point this out.

dra27 reviewed (commented):
In these `autoconf` days, it would be good if we were heading towards `./configure && make all && make install` - i.e. no more `world`, `opt`, `opt.opt`, `world.opt`, etc. but `configure` properly determining what's to be built and then `make` just doing it. That would certainly make this change simpler, but it's a very clear win - modulo the fix I note.

dra27 commented on Makefile:
```diff
@@ -679,9 +685,10 @@ endif
 	    "$(INSTALL_COMPLIBDIR)"
-	if test -n "$(WITH_OCAMLDOC)"; then \
-	  $(MAKE) -C ocamldoc installopt; \
 	fi
```
```suggestion
```

gasche replied:
Fixed, thanks!

gasche committed:
explanatory comment on when 'ocamlrun ocamldoc' fails

gasche commented:
@dra27, do you approve of this change? The root-Makefile part of the PR conflicts somewhat with #8839, and I think this one is "more ready", so I would be interested in building on top of it instead of in parallel.

dra27 commented:
I do entirely approve, but I’d like Sébastien to have a chance to comment, if that’s ok?

gasche commented:
That's fine with me, thanks.

(Re. `make all && make install`: it seems like a fine goal, but I'm not volunteering to do the restructuration work. I don't think this PR or the sucessor to #8839 would make it any easier or more difficult, given that they are not invasive -- they don't change the Makefile structure.)

xavierleroy commented:
Just to annoy everyone: what happens with platforms where ocamlopt is not supported?  There was a time when Debian (people who care about man pages) cared about hppa, m68k, and ia64, while ocamlopt support for these platforms was removed.  It was important for them that a fully-functional, bytecode-only OCaml system could be built.  In all honesty, I don't know if they or anyone else still care.

gasche commented:
If `ocamldoc.opt` is not present, then it is not used -- see the logic below.

```
ifeq "$(wildcard $(OCAMLDOC_OPT))" ""
  OCAMLDOC_RUN=$(OCAMLDOC_RUN_BYTE)
else
  OCAMLDOC_RUN=$(OCAMLDOC_RUN_OPT)
endif
```

I'm industrializing this technique in #8840.

gasche commented:
(cc @shindere)

shindere commented:
Sorry for the delay on this.

@gache: would you mind fixing the confilict in Changes, rebase and
perhaps cleaning up the history a bit?

I would then be happy to approve/merge.

gasche commented:
Thanks @shindere; I rebased and updated the PR.

Most commits are independent of each other and have their specific purpose, but I merged the Changes-entry commit into the main commit.

shindere commented:
Thanks, @gasche.

Waiting for CI to be happy and then either I can merge or any body else
includinng you can do so.

shindere commented:
David Allsopp (2019/08/30 05:42 -0700):
> Merged #8837 into trunk.

Thanks a lot @dra27 and thanks for having suggeted to let me review the
code!
